### PR TITLE
[TGC] Adds comprehensive rules to the TGC + a new starter guide

### PR DESCRIPTION
**Pull request by EuSouAFazer**
## About The Pull Request

This PR adds a Comprehensive Rules and a Starter Guide on the files of the repo. The idea is that these text files will get a wiki mirror so that you can read them online and in-game, alongside any changes to them automatically updating the manual in-game and on the wiki.

Relevant major rule changes from the current game:
1.2: Decks are 40 cards with a max of 4 cards of the same name rather than 30 cards with 3 of each.
1.8 & 3.6: The player that goes second can gain +1 mana on any turn, once per game. The game didn't had a mechanic to compensate for first player advantage so this is here for that.
**2:** The stack has been entirely removed. Quick/instant effects are still in the game however you cannot use cards in response to another card and before it resolves, only after it resolves. 
**3.4.2:** You can only activate an effect that has a specific trigger if that trigger has happened within the last Closed gamestate and now.

2 and 3.4.2 have a very special interaction that is very important for rulings going forward. Essentially what this does is so that if you have two cards that say "When you draw a card: Deal 1 damage to your opponent" and you draw a card you can only deal 1 damage to your opponent. This is because you draw a card; now the last thing that happened is "you drew a card", and thus you can activate the first card, dealing 1 damage to your opponent. Now the last thing that happened is "you dealt 1 damage to your opponent", meaning you can't use the second card. This has big and widespread implications, specially considering the turn player has priority when activating effects:

- If a creature with "When this destroys an enemy creature:" destroys an enemy creature that says "On Destroyed:" the opponent does not get to use their On Destroyed effect.
- During your turn (and your turn only) you can prevent your opponents' activated effects from happening by using an Instant card right when the opponents' trigger effect would happen.
- If you have multiple cards that trigger at the start of the battle phase you'd have to choose 1 to use and ignore the rest.

Notably, On turn start and On turn end don't suffer from this problem. This is because they can be set to trigger "On the standby phase:" and "On the end phase:" and only check to see if it's the current phase rather than for a specific trigger. This means effects that do should trigger every turn can safely be put in these triggers and not be skipped.

## Why It's Good For The Game
Why it's good to have the new rules: 

Yeah, I'm aware these rules are a bit of a bummer; a lot of triggers will get skipped and lead to a few feels bad moments. However, there are some  very very good reasons why I'm doing this:

- TGC is open source and infinitely expanding as people add new cards. This means that we need a rule system that can handle this expansion with as few rule nightmares as possible. It does increase the learning floor of the game however it also reduces by a very significant amount the level of ruling legalese that you need to know in order to play the game.
- Once you get the gist of the rule it makes it very clear when a trigger can work or not. This is not true for every card game - Yu-Gi-Oh and Magic The Gathering have some very, very janky interactions that are difficult to explain and wrap your head around because of not having a system like this.
- Not letting you use effects in response (only after) the opponent plays a card also prevents many ruling confusions. For example: If I use "Choose a creature; return it to the hand" and my opponent uses "Destroy a creature" in response destroying it before my card effect resolves, does it return it from the discard pile to my hand or does the effect does nothing? Furthermore, if I have a card that says "Give one of your creatures +1/+1 then draw a card" do I still draw if my opponent destroys all my creatures before the effect resolves? By unallowing players to activate effects between paying the mana cost of a card and it resolving, alongside adding a rule that states that in order to activate an effect you must be able to perform all of the effect (and not just parts of the effect) then we completely sidestep all of these issues.
- In sum - it lets us have a much, MUCH easier time later on when we have several different triggers and need to make rules on how to solve complex boardstates and avoid unintuitive rulings besides the first one. It's a speedbump to stop a car crash, essentially.

Why it's good to have rules & the starter guide on the repo:
- Having it on the wiki is bad as it means anyone can change the rules of the game without going through a PR or any sort of review - which is bad.
- Having it on a hackmd or a protected wiki page means that people cannot easily contribute to it and update it to changes to the game.
- By having it on the repo then mirroring it onto the wiki we have the best of both worlds. Anyone can easily edit it by making a PR however it still needs to go through review and get approved first. Furthermore, it means anyone adding new cards that require a rule change - for example, if someone wants to add a new type of card - can change the rules and add the cards in the same PR, saving headaches and streamlining the process.

## Changelog
:cl:
qol: [TGC] A new starter guide with updated info was made.
refactor: [TGC] The game's rules were completely updated with a comprehensive rules doc on the repo itself rather than on the wiki.
refactor: [TGC] Completely changed how the back end of the TGC rules work - the Stack was removed and replaced with a Open/Closed gamestate system. 
/:cl:

### CODE REVIEW

**Comment by EuSouAFazer** [comment]
PR also relevant to this one in that it helps make the cardpool work better with these changes (but neither is strictly needed for the other to work): https://github.com/tgstation/tgstation/pull/86288

**Comment by LemonInTheDark** [comment]
No comment on concept, I will say I don't think this is important enough to worry about people editing it, the wiki would be a fine place for this.

Also if you're pring text here it should be in markdown format.

**Comment by EuSouAFazer** [comment]
> No comment on concept, I will say I don't think this is important enough to worry about people editing it, the wiki would be a fine place for this.
> 
> Also if you're pring text here it should be in markdown format. 

Will change to markdown when I get home o7

As for why here and not on the wiki: If someone PRs a new set with cards that need to be explained in the rules (a new card type, for example) then having the rule change also be in the same PR is very very nice. It means maintainers can review the wording of the rule before it goes live, there wouldn't be any in-between time between the PR being accepted and the wiki being updated, etc.

It also means that if you just want to update the rules (such as tweaking it so a specific strategy gets less good, adjusting wording so that it's clearer what the rule is, refactoring rulings so the game is less messy etc) would still factor into the GBP economy rather than be outside of it. 

[It also prevents people from editing the wiki mid match to gain an advantage and other uncool angleshoots]

**Comment by EuSouAFazer** [comment]
I am having a lot more problems with markdown than what is reasonable, once I figure how to do line breaks we'll be back

**Comment by EuSouAFazer** [comment]
> will the maximum deck size be reflected with this new rule?

What do you mean?

**Comment by deathrobotpunch** [comment]
> What do you mean?

mechanically a TGC deck can only hold 30 cards

**Comment by EuSouAFazer** [comment]
> > What do you mean?
> 
> mechanically a TGC deck can only hold 30 cards

Huh, didn't knew that. Very weird - I think it's the only TGC mechanic that is actually coded in, outside of the holodeck? It's hardcoded too. Regardless, fixed, thanks for the catch!

**Comment by ExcessiveUseOfCobblestone** [comment]
> As for why here and not on the wiki: If someone PRs a new set with cards that need to be explained in the rules (a new card type, for example) then having the rule change also be in the same PR is very very nice. It means maintainers can review the wording of the rule before it goes live, there wouldn't be any in-between time between the PR being accepted and the wiki being updated, etc.

I dont think this is in scope, the maintainer team really shouldnt care to the point of halting a PR because you havent updated the rules in this file, which now if it gets missed also requires a PR and 24 hours minimum to be seen versus just fixing it on the wiki.

> 
> It also means that if you just want to update the rules (such as tweaking it so a specific strategy gets less good, adjusting wording so that it's clearer what the rule is, refactoring rulings so the game is less messy etc) would still factor into the GBP economy rather than be outside of it.

No gain here, wording of the rules for a card game that is pretty minor in terms of the overall game (not even related to job content) seems like it has no business giving (or taking) GBP.

Im not going to assume malice here but this PR alone claims it warrants refactor-level GBP over this when literally all you did was copy paste onto the repo. You didnt DO anything that made a system functionally better.

> 
> [It also prevents people from editing the wiki mid match to gain an advantage and other uncool angleshoots]

There is a history with edits, and someone touching something like the wiki to make a cheap attempt to trick someone seems like an admin issue akin to editing space law for the same reason.

Realistically speaking the feature is incredibly minor and I think this escalates the system to something much more serious than it should ever be in it's current iteration. Especially true if people want to include variants or homebrewed rules, that should definitely be something to be explored and should not be done through the PR process to be visible to others.

File: code/game/objects/items/tcg/tcg.dm [hunk 1]
```diff
@@ -139,7 +139,7 @@ GLOBAL_LIST_EMPTY(tcgcard_radial_choices)
 		user.put_in_hands(new_deck)
 	if(istype(item, /obj/item/tcgcard_deck))
 		var/obj/item/tcgcard_deck/old_deck = item
-		if(length(old_deck.contents) >= 30)
```
Comment (ExcessiveUseOfCobblestone) [inline code/game/objects/items/tcg/tcg.dm:142]: this needs to be a define

File: code/game/objects/items/tcg/tcg.dm [hunk 2]
```diff
@@ -256,7 +256,7 @@ GLOBAL_LIST_EMPTY(tcgcard_radial_choices)
 /obj/item/tcgcard_deck/attackby(obj/item/item, mob/living/user, params)
 	. = ..()
 	if(istype(item, /obj/item/tcgcard))
-		if(contents.len >= 30)
```
Comment (ExcessiveUseOfCobblestone) [inline code/game/objects/items/tcg/tcg.dm:259]: above

**Comment by EuSouAFazer** [comment]
Most of this is getting moved to https://github.com/tgstation/tgc-rules/pull/1
Gonna re-make this PR with just the change to the max decksize becoming a define, no other changes here

**Pull request closed without merging**